README
This is the BIP IRC Proxy README.

Bip can be used in two different ways:
- Old school bnc user style: easy and straightforward.
- Unix service style with and init.d scripts and the logs in /var/log

This small README file explains the usage "Old school" with which :
 - you do not need the root privileges.
 - gives easy access to the logs to the owner of the shell.

Table of contents :

	I.	Installation
	II.	Configuration
	   A.	  Manual configuration
	   B.	  Automated configuration
	III.	Running bip
	IV.	Using bip
	   A.	  Connecting your client(s)
	   B.	  Backlog and flood control
	   C.	  Multiple users and ident issues



I. INSTALLATION

    Install bip on the machine that will be running bip (which is likely to be
    your personnal or shared server) either compiling the package or using your
    distro's package. Then create a configuration file.

    Choose your distribution package if available. If not, build bip the
    old-fashioned way. You will need make, gcc, lex, yacc, automake,
    autoconf-archive and optionally libssl-dev to build bip.

    Just issue:

    From bip-X.Y.Z.tar.gz package:
    # ./configure && make

    From repository:
    #  autoreconf -i
    # ./configure --enable-maintainer-mode && make

    If openssl and its developement files are installed, bip should build with
    SSL support. After a successful build the bip binary can be found in
    ./src/bip.

    By default, "-Werror" is used. If you encounter warnings, you could try:
    # CFLAGS="-Wno-error" ./configure && make


II. CONFIGURATION

    First of all, create your bip configuration an log directory:

    # mkdir -p ~/.bip/logs

    There are two ways to create your bip configuration :
    - edit the sample bip.conf file to match your needs
    - use the bipgenconfig script to easily generate a configuration

    If you want to connect to bip using an SSL client, you'll need to create
    a certificate / key pair (in a bip.pem file) to allow bip to serve SSL
    sockets.

    A. MANUAL CONFIGURATION

      If you are using a distribution package, the bip.conf sample configuration
      file is likely to be shipped in /usr/share/doc/bip/examples/bip.conf.gz or
      something similar.

      If not, you'll find sample configuration file in the source package's
      `samples' subdirectory.

      Put the uncompressed configuration file in your ~/.bip directory (its
      path should be ~/.bip/bip.conf), and edit it, most importantly the "user"
      section that contains information about you and the servers you will want
      to connect to. The "name" field in the "user" section is your login to
      connect to bip.
      The "name" field of the "connection" subsections are the server identifier
      for when you connect to bip.

      The "password" field is a hash of the password you will use to connect to
      bip. To generate a hash value from a password, use bipmkpw, program which
      comes in the bip package and source.

      If you've set client_side_ssl to true, you'll need to generate a bip.pem
      file containing a certificate / key pair. In order to do so, you can use
      the third party `openssl' binary :

      # openssl req -new -x509 -days 365 -nodes -out bip.pem -keyout bip.pem

      You can then remove the passphrase with :

      # openssl x509 -subject -dates -fingerprint -noout -in bip.pem

    B. AUTOMATED CONFIGURATION

      You can also use the bipgenconfig script to generate a new configuration.
      This script will also help you generate the SSL certificate / key pair
      needed for clients to connect to BIP through SSL.

      This script can be found either in the source package's `scripts'
      directory or shipped with your distribution's package.

      Using the script is very simple, and it will generate a configuration
      file but won't overwrite any existing configuration.

      It will ask you the path to the bipmkpw binary, to automatically hash the
      passwords you'll provide. Please make sure to enter the correct path to
      the binary or you might observe unexpected behaviour.

      You'll need to move the generated configuration from bip.conf.autogen to
      bip.conf and the generated PEM file from bip.pem.autogen to bip.pem (or
      whatever path you've configured in bip.conf).


III. RUNNING BIP

    Once all this is configured, start bip as your regular user:

    # ./src/bip

    If you have installed bip in your path (or if you are using you
    distribution's package), simply use:

    # bip

    Once bip starts, it connects to the different servers your defined in
    all "user"'s "connection" blocks.


IV. USING BIP

    A. CONNECTING YOUR CLIENT(S)

      Then you want to use your regular irc client and connect to bip.  Point
      your client to the machine bip is running and set the proper port number
      (defined in your bip.conf). You should then configure the client to use a
      specific irc server password constructed this way:
  
      user:password:connection
  
      The user is the name field of the "user" section, the password is the
      password (*not* the hash) corresponding to the "password" field of the
      same user section (which is the hash generated with bipmkpw) and the
      connection is the "name" field of the "connection" subsection. This is
      how bip authenticates you and puts your client to the correct network.
  
      Using the default (or sample file) configuration, logs are in ~/.bip/logs/
  
    B. BACKLOG AND FLOOD CONTROL

      Bip has a backlogging system which will send back parts of the last logs
      upon client connection. Depending on your configuration, that may mean a
      *lot* of data sent back to your client.
      
      Users' messages will be replayed as if they were being sent at the moment
      your client connects to bip, and if not disabled, system messages will
      appear as coming from the "-bip" user.
      
      Considering that, you may want to disable your client's anti-flood system,
      totally or not, depending on it's flexibility.
      Since bip doesn't replay CTCP messages, you can safely let your client's
      anti-flood system manage them.
      
      [Xchat]
      If you're using Xchat, you can "disable" it by issuing these commands :
      /set flood_msg_num = 1000
      /set flood_msg_time = 10
      In fact you'll tell xchat to activate its anti-flood system when you're
      receiving more than 1000 messages in less than 10 seconds.
      
      If you forgot to set these, private messages may not appear in separate
      tabs as usual. If so, simply issue a :
      /set gui_auto_open_dialog on

    C. MULTIPLE USERS AND IDENT ISSUES

      When you host many connections to the same IRC network, you might have
      more connections than allowed by the network from one host.

      Depending on the network and the services it runs, session limits may be
      enforced either matching only your ip address/hostname, or matching the
      username/ident part too.

      To avoid being killed for session limit exceeded, you should define a
      default_username in each user {}; block. A user without default_username
      would appear as ~bip@yourhost if bip is the system user running bip. With
      a default_username set to "myuser", he would appear as ~myuser@yourhost,
      which may be sufficient for most networks.

      If the network you're on is a bit more demanding, you can set up an
      oidentd server on your host, and enable oidentd spoofing support
      ('write_oidentd = true;' option in bip configuration file). Let's
      say bip is the system user running bip, you should add to
      your /etc/oidentd.conf :

        user "bip" {
            default {
                allow spoof_all
                allow spoof_privport
                allow spoof
            }
        }

      Then reload oidentd and make sure that ~bip is accessible (+rx) by the
      user running oidentd (which means most of the time ~bip should be world
      readable and browsable +rx).

      If you already have a ~bip/.oidentd.conf file, don't worry, bip'll only
      add its entries without deleting any of the contents of the file.
      This step should remove the "~" character from the username/ident part of
      your ircmask, and thus satisfy some networks.

      If the network is still killing you for session limit exceeded, you'll
      have to contact it's admins and ask them for an exception on your host or
      ip address.


  Happy ircing!

-- Arnaud Cornet <nohar@t1r.net> and Loïc Gomez <opensource@kyoshiro.org>



NEWS
2022-03-10 (0.9.3) "sanitize our bip and add some layers"
- Add SASL authentication support (EXTERNAL, PLAIN)
- Handle some recommended GCC flags
- fix /bip user info command

2021-10-12 (0.9.2) "for the love of unsubsquirrel"
- Update and fix minor issues in sample bip and vim configuration files

2021-10-12 (0.9.1) "one more"
- ENABLED might not be defined: use a default value.

2021-10-08 (0.9.0) "it's never too late"
Visible changes:
- "backlog_no_timestamp" is deprecated, "backlog_timestamp" should be used
  instead. The allowed values for this new parameter are: "none", "time",
  "datetime". "time" is the default value and allow to keep the default
  behavior.
- allow to set TLS ciphers: for bip client side (client_side_ciphers),
  globally for all connections (ssl_default_ciphers) and foreach connection
  (ssl_ciphers).
- allow to set DH parameter used for clients SSL connections with the new
  client_side_dh_param parameter.
- Add write_oidentd_file parameter. Existing oidentd users must use
  'write_oidentd = true' in their config to keep existing behavior.

2013-10-19 (0.8.9) "If it's really that urgent, why don't you do it yourself?"
- code fixes
- buildsys fixes (bison 2.6 compatibility and others)
- improved TRUST OK command replies when there's no untrusted certificate
- fixed GCC warnings

2011-04-14 (0.8.8) "spring release"
- Allow to disable logs by connection.
- When global option log is disabled, query are not backlogged.
- Avoid segfault when global option log is false.

2011-01-18 (0.8.7) "bazooka"
- Fix "FATAL: list_remove: item not found"
- Fix build errors on armel
- minor fixes

2010-09-12 (0.8.6) "CVE-2010-3071"
- Fix for CVE-2010-3071.

2010-08-07 (0.8.5) "Unpacking"
- Now builds position independant executables

2009-11-15 (0.8.4) "I’ll have the same thing please"
- Fix build OpenSSL detection issue.

2009-11-15 (0.8.3) "Hey Dublin! What’s the craic?"
- Fixes a fatal() on gamesurge networks.

2009-08-24 (0.8.2) "Farewell Paris"
- Fix hanging bip on hanging client connexions.

2009-07-17 (0.8.1) "Ran ran ru in Praha"
- Fixed an annoying bug that resulted in too much backlog.
- Fix TRUST command.
- Close some unused log files.

2009-03-02 (0.8.0) "swelling millionnaire"
- Cosmetic bug fixes since last rc.

2009-02-02 (0.8.0-rc1) "got booze?"
- One bug fixed and a client hack added to avoid the "+" or "-" prefix in
  irssi when connecting irssi and xchat to the same bip connection.

2009-01-24 (0.8.x)
Visible changes:
- One logfile per nick for queries, instead of the messy privates.x.log
  Slight log format change (now the format of the log of queries is very
  similar to the on used for channels)
- "Window"-local blreset and blreset on talk. By default the blreset_on_talk
  now only clears the backlog of the query you talked to or the channel you
  talked to.  The /quote bip blreset command now can take a query name or a
  channel name as an argument. "/quote bip blreset" still resets the backlog of
  the whole network.
- /quote bip backlog x command where x is the number of hours will backlog x
  hours.
- autorejoin on kick by default. Can be disabled with the option
  autojoin_on_kick = false in a connection block.
- bip SSL on the client<->bip part now support Ephemeral Diffie Hellman key
  exchange.

2008-10-24 (0.7.5) "But I have promises to keep, And miles to go before I
 sleep, And miles to go before I sleep."
- Halfop handling fixed, and a shameful segfault. The segfault may be
  exploitable by clients, but only after successful login. Please upgrade.

2008-06-28 (0.7.4) "But where is 0.7.3 ?!"
- Some usefull bugfixes, especially for unreal irc network users.

2008-04-05 (0.7.2) "Enjoy the fish"
- Contains a compilation fix for OpenBSD.

2008-04-02: bip 0.7.1 "hot mama"
- Contains a configuration validation fix, and a memory leak fix.

2008-02-16 (0.7.0) "birthday party"
- The good stuff: sighup support, /bip reload should also work now
- New commands (allow a user to add a new network without restarting bip for
  instance)
- Better user feedback when issuing /bip commands.
- Lots of new backlog options, some of which can now be set per user instead of
  globally.
- And a few fixes on top of that.

2008-02-07  Arnaud Cornet <nohar@t1r.net>
What's to expect in bip now:
- Support kill -HUP and /bip reload cleanly.
- Close/reopen files on sighup.
- Have some reply when running BIP commands.  ssl_check_mode, always_backlog,
  backlog, bl_msg_only, backlog_lines, backlog_no_timestamp, blreset_on_talk be
  settable at the connection block rather than the user block.
- Add backlog option to channels. Useful if you don't care about one particular
  channel.
- bip_use_notice: Add this option to receive bip message through notices.
- Make SSL "basic" certificate check more loose (more SSH-like).  In "basic"
  mode, normally un trusted certificates get to be trusted if they manually
  trusted by user. This provides an SSH like private key auth mechani Expired
  certificates were considered invalid in this mode which does not make much
  sense.
- blreset: add -q option to /bip blreset command if you don't want it to reply
  anything (for scripts).
- Support add_conn and del_conn bip commands for live connection addition and
  removal.
- Have a bold arrow to highligh your own word on private messages and multiple
  connected clients.

2007-12-09
- Bip has now two types of users "admins" and normal user. Some commands are
  reserved to admin. Add a admin = true; statement in your config (in user
  block) accordingly.

2007-10-29
- Certificate validation change. In "basic" mode, expired certificates are now
  accepted as long as they are in store (therefore trusted). This makes the
  basic mode be more SSH like. Some extreme security zealots might want to be
  warned.

2007-09-02
- as of now log parameters go in the user {} statment. This breaks
  every config and there is no backwrads compatibility as of now.
- Lots of internal changes, expect crashes.


src/util.c
/*
 * $Id: util.c,v 1.60 2005/04/12 19:34:35 nohar Exp $
 *
 * This file is part of the bip project
 * Copyright (C) 2004,2005 Arnaud Cornet
 * Copyright (C) 2004,2005,2022 Loïc Gomez
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * See the file "COPYING" for the exact licensing terms.
 */

#include "config.h"
#include "connection.h"
#include "util.h"
#include <stdarg.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>

extern int conf_log_level;
extern int conf_log_system;
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wstrict-prototypes"
extern int errno;
#pragma GCC diagnostic pop
extern FILE *conf_global_log_file;

void memory_fatal(void)
{
	fflush(conf_global_log_file);
#define OOMMSG "Out of memory.\n"
	fwrite(OOMMSG, (size_t)1, strlen(OOMMSG), conf_global_log_file);
#undef OOMMSG
	fflush(conf_global_log_file);
	exit(28);
}

void *bip_malloc(size_t size)
{
	void *r;

	assert(size < INT_MAX / 4);

	r = malloc(size);
	if (!r)
		memory_fatal();
	return r;
}

void *bip_calloc(size_t nmemb, size_t size)
{
	void *r = calloc(nmemb, size);
	if (!r)
		memory_fatal();
	return r;
}

void *bip_realloc(void *ptr, size_t size)
{
	void *r;

	assert(size < INT_MAX / 4);

	r = realloc(ptr, size);
	if (size > 0 && r == NULL)
		memory_fatal();
	return r;
}

void bip_cfree(const void *ptr)
{
	if (!ptr)
		return;
// there's no other way to free a const pointer
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wcast-qual"
	free((void *)ptr);
#pragma GCC diagnostic pop
}

char *bip_strdup(const char *str)
{
	char *r = strdup(str);
	if (!r)
		memory_fatal();
	return r;
}

char *bip_strcat_fit(size_t *remaining, char *str, const char *str2)
{
	char *res;

	if (!remaining || !str || !str2) {
		mylog(LOG_DEBUGVERB, "bip_strcat_fit: got NULL pointer");
		return NULL;
	}

	res = memccpy(str, str2, '\0', *remaining);
	if (!res) {
		mylog(LOG_DEBUGTOOMUCH,
		      "bip_strcat_fit: memccpy() failed, remaining %lu",
		      *remaining);
		return NULL;
	}

	res--;
	if (res < str) {
		mylog(LOG_DEBUG, "bip_strcat_fit: memccpy res < str");
		return NULL;
	}
	(*remaining) -= (size_t)(res - str);
	return res;
}

#define STRCATF_BUF_MAXLEN 1024
char *bip_strcatf_fit(size_t *remaining, char *str, const char *fmt, ...)
{
	va_list ap;
	char str2[STRCATF_BUF_MAXLEN + 1];
	int written;
	char *res = NULL;

	if (!remaining || !str || !fmt) {
		mylog(LOG_DEBUGVERB, "bip_strcatf_fit: NULL pointer");
		return NULL;
	}

	if (*remaining > STRCATF_BUF_MAXLEN) {
		mylog(LOG_ERROR,
		      "bip_strcatf_fit: remaining "
		      "is over STRCATF_BUF_MAXLEN");
	}

	va_start(ap, fmt);
	str2[*remaining] = '\0';
	written = vsnprintf(str2, *remaining, fmt, ap);
	if (written < 0) {
		mylog(LOG_ERROR, "bip_strcatf_fit: vsnprintf failed with: %s",
		      strerror(errno));
		return NULL;
	}

	if ((unsigned)written >= *remaining) {
		mylog(LOG_DEBUGVERB,
		      "bip_strcatf_fit,vsnprintf: no space left");
		goto end;
	}

	res = memccpy(str, str2, '\0', *remaining);
	if (!res) {
		mylog(LOG_DEBUGTOOMUCH,
		      "bip_strcatf_fit: memccpy() failed, "
		      "remaining %lu",
		      *remaining);
		goto end;
	}

	if (res < str) {
		mylog(LOG_DEBUG, "bip_strcatf_fit: memccpy res < str");
		goto end;
	}
	res--;
	(*remaining) -= (size_t)(res - str);

end:
	va_end(ap);
	return res;
}

void bip_clock_gettime(clockid_t clockid, struct timespec *tp)
{
	int err = clock_gettime(clockid, tp);

	if (err != 0)
		fatal("clock_gettime: %s", strerror(errno));

	if (tp->tv_sec < 0 || tp->tv_nsec < 0)
		fatal("clock_gettime returned negative time");
}

/*
 * <nick> ::= <letter> { <letter> | <number> | <special> }
 * <special> ::= '-' | '[' | ']' | '\' | '`' | '^' | '{' | '}'
 */
int is_valid_nick(char *str)
{
	char *tmp;
	if (!str || !isalpha(*str))
		return 0;

	tmp = str;
	while (*tmp != '\0'
	       && (isalnum(*tmp) || *tmp == '-' || *tmp == '[' || *tmp == ']'
		   || *tmp == '\\' || *tmp == '`' || *tmp == '^' || *tmp == '{'
		   || *tmp == '}' || *tmp == '|' || *tmp == '_'))
		tmp++;
	return (*tmp == '\0');
}

int is_valid_username(char *str)
{
	char *tmp;
	if (!str || *str == '\0' || *str == ' ' || *str == '\n' || *str == '\r')
		return 0;

	tmp = str;
	while (*tmp != '\0' && *tmp != ' ' && *tmp != '\0' && *tmp != '\r'
	       && *tmp != '\n')
		tmp++;
	return (*tmp == '\0');
}

char *timestamp(void)
{
	static char ts[20];
	time_t tv;
	struct tm *tm;

	time(&tv);
	tm = localtime(&tv);

	strftime(ts, (size_t)20, "%d-%m-%Y %H:%M:%S", tm);
	return ts;
}

char *hrtime(time_t s)
{
	static char ts[20];
	struct tm *tm;

	if (s == 0)
		return "never";
	tm = localtime(&s);

	strftime(ts, (size_t)20, "%d-%m-%Y %H:%M:%S", tm);
	return ts;
}

#ifdef HAVE_LIBSSL
char *checkmode2text(int v)
{
	switch (v) {
	case SSL_CHECK_BASIC:
		return "basic";
	case SSL_CHECK_CA:
		return "ca";
	default:
		return "none";
	}
}
#endif

void _mylog(int level, char *fmt, va_list ap)
{
	char *prefix;

	if (!conf_log_system)
		return;

	if (level > conf_log_level)
		return;

	switch (level) {
	case LOG_FATAL:
		prefix = "FATAL: ";
		break;
	case LOG_DEBUGVERB:
		prefix = "DEBUGVERB: ";
		break;
	case LOG_DEBUG:
		prefix = "DEBUG: ";
		break;
	case LOG_ERROR:
		prefix = "ERROR: ";
		break;
	case LOG_WARN:
		prefix = "WARNING: ";
		break;
	case LOG_INFO:
		prefix = "";
		break;
	default:
		prefix = "";
		break;
	}

	fprintf(conf_global_log_file, "%s %s", timestamp(), prefix);
	vfprintf(conf_global_log_file, fmt, ap);
	fprintf(conf_global_log_file, "\n");
#ifdef DEBUG
	fflush(conf_global_log_file);
#endif
}

void mylog(int level, char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	_mylog(level, fmt, ap);
	va_end(ap);
}

#ifdef HAVE_BACKTRACE
#include <execinfo.h>
void dump_trace(void)
{
	void *array[32];
	int size;

	size = backtrace(array, 32);
	fflush(conf_global_log_file);
	backtrace_symbols_fd(array, size, fileno(conf_global_log_file));
}
#endif

extern char *conf_pid_file;
void fatal(char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	_mylog(LOG_FATAL, fmt, ap);
	va_end(ap);

#ifdef HAVE_BACKTRACE
	dump_trace();
#endif

	exit(200);
}

/*
 * list handling functions
 */

int list_ptr_cmp(const void *a, const void *b)
{
	if (a == b)
		return 0;
	return -1;
}

void list_init(list_t *l, int (*cmp)(const void *, const void *))
{
	l->first = NULL;
	l->last = NULL;
	l->cmp = cmp;
}

list_t *list_new(int (*cmp)(const void *, const void *))
{
	list_t *l;
	l = bip_malloc(sizeof(list_t));
	list_init(l, cmp);
	return l;
}

static struct list_item *list_item(void *ptr)
{
	struct list_item *l;
	l = bip_malloc(sizeof(struct list_item));
	l->ptr = ptr;
	l->next = NULL;
	l->prev = NULL;
	return l;
}

void list_add_first(list_t *list, void *ptr)
{
	struct list_item *li;

	assert(list);
	if (!ptr)
		fatal("Cannot add NULL ptr to list.");
	li = list_item(ptr);
	if (!list->first) {
		list->first = list->last = li;
		return;
	}
	li->next = list->first;
	list->first = li;
	li->next->prev = li;
}

void list_add_first_uniq(list_t *list, void *ptr)
{
	assert(list);
	assert_msg(ptr, "Cannot add NULL ptr to list.");
	if (list_get(list, ptr))
		return;
	list_add_first(list, ptr);
}

void list_add_last(list_t *list, void *ptr)
{
	struct list_item *li;

	assert(list);
	assert_msg(ptr, "Cannot add NULL ptr to list.");
	li = list_item(ptr);
	if (!list->first) {
		list->first = list->last = li;
		return;
	}
	li->prev = list->last;
	list->last = li;
	li->prev->next = li;
}

void *list_get_first(list_t *list)
{
	assert(list);
	if (!list->first)
		return NULL;
	return list->first->ptr;
}

void *list_get_last(list_t *list)
{
	assert(list);
	if (!list->last)
		return NULL;
	return list->last->ptr;
}

void *list_remove_first(list_t *list)
{
	struct list_item *l;
	void *ptr;

	assert(list);
	ptr = list_get_first(list);
	if (!ptr)
		return NULL;
	l = list->first;
	list->first = list->first->next;
	if (list->first == NULL)
		list->last = NULL;
	free(l);
	return ptr;
}

void *list_remove_last(list_t *list)
{
	struct list_item *l;
	void *ptr;

	assert(list);
	ptr = list_get_last(list);

	if (!ptr)
		return NULL;
	l = list->last;
	list->last = list->last->prev;
	if (list->last == NULL)
		list->first = NULL;
	free(l);
	return ptr;
}

void *list_remove_if_exists(list_t *list, const void *ptr)
{
	list_iterator_t li;
	int debug = 0;
	void *ret = 0;

	assert(list);
	assert_msg(list->cmp, "list does not have a cmp function");

	for (list_it_init(list, &li); list_it_item(&li); list_it_next(&li)) {
		if (list->cmp(list_it_item(&li), ptr) == 0) {
			if (debug == 1)
				fatal("%x appears twice in list\n", ptr);
			ret = list_it_remove(&li);
			debug = 1;
		}
	}
	if (debug)
		return ret;
	return NULL;
}

void *list_remove(list_t *list, const void *ptr)
{
	void *ret;
	if (!(ret = list_remove_if_exists(list, ptr)))
		fatal("list_remove: item not found");
	return ret;
}

void *list_get(list_t *list, const void *ptr)
{
	struct list_item *it;

	assert_msg(list->cmp, "list_get: list does not have a cmp function");

	for (it = list->first; it; it = it->next) {
		if (list->cmp(it->ptr, ptr) == 0)
			return it->ptr;
	}
	return NULL;
}

int list_is_empty(list_t *l)
{
	assert(l);
	return (l->first ? 0 : 1);
}

void list_it_init(list_t *list, list_iterator_t *ti)
{
	assert(list && ti);
	ti->list = list;
	ti->cur = list->first;
	ti->next = NULL;
}

void list_it_init_last(list_t *list, list_iterator_t *ti)
{
	assert(list && ti);
	ti->list = list;
	ti->cur = list->last;
	ti->next = NULL;
}

void *list_it_remove(list_iterator_t *li)
{
	assert(li);
	if (!li->cur)
		return NULL;

	if (li->cur->prev)
		li->cur->prev->next = li->cur->next;
	else
		li->list->first = li->cur->next;

	if (li->cur->next)
		li->cur->next->prev = li->cur->prev;
	else
		li->list->last = li->cur->prev;

	void *ptr = li->cur->ptr;
	struct list_item *item = li->cur;
	li->next = li->cur->next;
	li->cur = NULL;
	free(item);
	return ptr;
}

void list_free(list_t *t)
{
	assert(t);
	assert(list_is_empty(t));
	free(t);
}

void list_append(list_t *dest, list_t *src)
{
	assert(dest && src);
	if (src->last == NULL)
		return;
	if (dest->first == NULL) {
		dest->first = src->first;
		dest->last = src->last;
		src->first = src->last = NULL;
		return;
	}
	src->first->prev = dest->last;
	dest->last->next = src->first;
	dest->last = src->last;
	src->first = src->last = NULL;
}

/*
 * Hash stuff
 */

struct hash_item {
	char *key;
	void *item;
};

static int hash_item_nocase_cmp(const struct hash_item *a, const char *b)
{
	return strcasecmp(a->key, b);
}

static int hash_item_cmp(struct hash_item *a, char *b)
{
	return strcmp(a->key, b);
}

void hash_init(hash_t *h, int options)
{
	int i;

	assert(h);
	memset(h, 0, sizeof(hash_t));
	for (i = 0; i < 256; i++) {
		switch (options) {
		case HASH_NOCASE:
			list_init(&h->lists[i],
				  (int (*)(const void *,
					   const void *))hash_item_nocase_cmp);
			break;
		case HASH_DEFAULT:
			list_init(&h->lists[i],
				  (int (*)(const void *,
					   const void *))hash_item_cmp);
			break;
		default:
			fatal("wrong hash option %d", options);
		}
	}
}

void hash_clean(hash_t *h)
{
	int i;
	struct hash_item *hi;

	assert(h);
	for (i = 0; i < 256; i++) {
		while ((hi = list_remove_first(&h->lists[i]))) {
			free(hi->key);
			free(hi);
		}
	}
}

void hash_free(hash_t *h)
{
	assert(h);
	hash_clean(h);
	free(h);
}

hash_t *hash_new(int options)
{
	hash_t *h;
	h = bip_malloc(sizeof(hash_t));
	hash_init(h, options);
	return h;
}

/* Now we have a real hash, but we use only the last byte of it :p */
static unsigned char hash_func(const char *pkey)
{
	char c;
	unsigned long hash = 5381; /* 5381 & 0xff makes more sense */

	// toupper should not return negative values (only char compatible int)
	while ((c = *pkey++))
		hash = ((hash << 5) + hash) ^ (long unsigned)toupper(c);
	return (unsigned char)hash;
}

void hash_insert(hash_t *hash, const char *key, void *ptr)
{
	struct hash_item *it;

	assert(hash && key);
	if (hash_get(hash, key))
		fatal("Element with key %s already in hash %x\n", key, hash);

	it = bip_malloc(sizeof(struct hash_item));
	it->key = bip_strdup(key);
	it->item = ptr;
	list_add_first(&hash->lists[hash_func(key)], it);
}

int hash_includes(hash_t *hash, const char *key)
{
	struct hash_item *hi;
	list_t *list;

	assert(hash && key);
	list = &hash->lists[hash_func(key)];
	hi = list_get(list, key);
	return hi != NULL;
}

void *hash_get(hash_t *hash, const char *key)
{
	struct hash_item *hi;
	list_t *list;

	assert(hash && key);
	list = &hash->lists[hash_func(key)];
	hi = list_get(list, key);
	if (!hi)
		return NULL;
	return hi->item;
}

void *hash_remove_if_exists(hash_t *hash, const char *key)
{
	assert(hash && key);
	if (hash_get(hash, key) == NULL)
		return NULL;
	return hash_remove(hash, key);
}

void *hash_remove(hash_t *hash, const char *key)
{
	struct hash_item *it;
	void *ptr;

	assert(hash && key);
	it = (struct hash_item *)list_remove(&hash->lists[hash_func(key)], key);
	if (!it)
		return NULL;
	ptr = it->item;
	free(it->key);
	free(it);
	return ptr;
}

int hash_is_empty(hash_t *h)
{
	int i;

	assert(h);
	for (i = 0; i < 256; i++) {
		if (!list_is_empty(&h->lists[i]))
			return 0;
	}
	return 1;
}

void hash_it_init(hash_t *h, hash_iterator_t *hi)
{
	assert(h);
	memset(hi, 0, sizeof(hash_iterator_t));
	hi->hash = h;

	while (hi->list < 256 && list_is_empty(&h->lists[hi->list]))
		hi->list++;
	if (hi->list < 256)
		list_it_init(&h->lists[hi->list], &hi->lit);
}

void hash_it_next(hash_iterator_t *hi)
{
	assert(hi);
	list_it_next(&hi->lit);
	if (!list_it_item(&hi->lit)) {
		do {
			hi->list++;
			if (hi->list == 256)
				return;
		} while (list_is_empty(&hi->hash->lists[hi->list]));
		list_it_init(&hi->hash->lists[hi->list], &hi->lit);
	}
}

void *hash_it_item(hash_iterator_t *h)
{
	struct hash_item *hi;

	assert(h);
	hi = list_it_item(&h->lit);
	if (!hi)
		return NULL;
	return hi->item;
}

const char *hash_it_key(hash_iterator_t *h)
{
	struct hash_item *hi;

	assert(h);
	hi = list_it_item(&h->lit);
	if (!hi)
		return NULL;
	return hi->key;
}

void *hash_it_remove(hash_iterator_t *hi)
{
	struct hash_item *hitem;
	void *ptr;

	assert(hi);
	hitem = list_it_remove(&hi->lit);

	ptr = hitem->item;
	free(hitem->key);
	free(hitem);
	return ptr;
}

void hash_dump(hash_t *h)
{
	hash_iterator_t it;
	assert(h);
	for (hash_it_init(h, &it); hash_it_key(&it); hash_it_next(&it))
		printf("%s => %p\n", hash_it_key(&it), hash_it_item(&it));
}

list_t *hash_keys(hash_t *hash)
{
	hash_iterator_t hi;
	list_t *ret;

	assert(hash);
	ret = list_new(NULL);

	for (hash_it_init(hash, &hi); hash_it_item(&hi); hash_it_next(&hi))
		list_add_last(ret, bip_strdup(hash_it_key(&hi)));

	return ret;
}

void hash_rename_key(hash_t *h, const char *oldk, const char *newk)
{
	assert(h && oldk && newk);
	if (strcmp(oldk, newk) == 0)
		return;
	hash_insert(h, newk, hash_remove(h, oldk));
}

char *bip_strmaydup(char *s)
{
	if (!s)
		return s;
	return bip_strdup(s);
}

void strucase(char *s)
{
	while (*s) {
		*s = (char)toupper(*s); // toupper, safe to cast to char
		s++;
	}
}

int ischannel(char p)
{
	return (p == '#' || p == '&' || p == '+' || p == '!');
}

void array_init(array_t *a)
{
	memset(a, 0, sizeof(array_t));
}

array_t *array_new(void)
{
	array_t *a;

	a = bip_malloc(sizeof(array_t));
	array_init(a);
	return a;
}

void array_ensure(array_t *a, int index)
{
	assert(a && index >= 0);

	if (array_includes(a, index))
		return;
	a->elemv = bip_realloc(a->elemv, sizeof(void *) * (size_t)(index + 1));
	// a->elemc should be lower than index + 1
	memset(a->elemv + a->elemc, 0,
	       sizeof(void *) * (size_t)(index + 1 - a->elemc));
	a->elemc = index + 1;
}

const void *array_drop(array_t *a, int index)
{
	int i;
	const void *ret;

	assert(a && array_includes(a, index));

	ret = a->elemv[index];
	for (i = index; i < array_count(a) - 1; i++)
		a->elemv[i] = a->elemv[i + 1];
	a->elemc--;
	return ret;
}

array_t *array_extract(array_t *a, int index, int upto)
{
	array_t *ret;
	int i;

	assert(a && array_includes(a, index));
	if (upto == -1)
		upto = a->elemc;
	assert((index == 0 && upto == 0) || array_includes(a, upto - 1));
	assert(index <= upto);

	ret = array_new();
	if (index == upto)
		return ret;

	/* here we have index < upto */
	array_ensure(ret, upto - index - 1);
	for (i = 0; i < ret->elemc; i++)
		ret->elemv[i] = a->elemv[i + index];
	return ret;
}

void array_deinit(array_t *a)
{
	assert(a);
	if (a->elemv)
		free(a->elemv);
	array_init(a);
}

void array_free(array_t *a)
{
	assert(a);
	if (a->elemv)
		free(a->elemv);
	free(a);
}
